pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java

```diff
@@ -656,6 +662,19 @@ public synchronized void reset() {
         flushRequired = false;
     }
 
+    public synchronized int compareTo(MessageId messageId) {
+        // The position of a message in the batch (BatchMessageIdImpl) must precede the batch itself (MessageIdImpl)
+        if (this.messageId instanceof BatchMessageIdImpl && (!(messageId instanceof BatchMessageIdImpl))) {
+            int result = ((BatchMessageIdImpl) this.messageId).toMessageIdImpl().compareTo(messageId);
```

Comment:
We can transform `messageId` to `BatchMessageIdImpl`, and `batchIndex = Integer.MAX_VALUE`

Reply:
I used `MessageIdImpl#messageIdCompare` instead, PTAL again.